src/index-format/granular_marshal.ml

```diff
@@ -0,0 +1,82 @@
+type store = in_channel
+
+type 'a link = 'a repr ref
+
+and 'a repr =
+  | Small of 'a
+  | Serialized of {loc: int}
+  | On_disk of {fd: store; loc: int; schema: 'a schema}
+  | In_memory of 'a
+
+and 'a schema = iter -> 'a -> unit
+
+and iter = {yield: 'a. 'a link -> 'a schema -> unit}
+
+let schema_no_sublinks : _ schema = fun _ _ -> ()
+
+let link v = ref (In_memory v)
+
+let fetch_loc fd loc schema =
+  seek_in fd loc ;
+  let v = Marshal.from_channel fd in
+  let rec iter =
+    { yield=
+        (fun lnk schema ->
+          match !lnk with
+          | Serialized {loc} -> lnk := On_disk {fd; loc; schema}
+          | Small v -> schema iter v; lnk := In_memory v
+          | _ -> assert false ) }
+  in
+  schema iter v ; v
+
+let fetch lnk =
+  match !lnk with
+  | In_memory v -> v
+  | Serialized _ | Small _ -> invalid_arg "fetch: serialized"
+  | On_disk {fd; loc; schema} ->
+    let v = fetch_loc fd loc schema in
+    lnk := In_memory v ;
+    v
+
+let ptr_size = 8
+
+let binstring_of_int v =
+  String.init ptr_size (fun i -> Char.chr ((v lsr i lsl 3) land 255))
+
+let int_of_binstring s =
+  Array.fold_right
+    (fun v acc -> (acc lsl 8) + v)
+    (Array.init ptr_size (fun i -> Char.code s.[i]))
+    0
+
+let write ?(flags = []) fd root_schema root_value =
+  let pt_root = pos_out fd in
+  output_string fd (String.make ptr_size '\000') ;
+  let rec iter =
+    { yield=
+        (fun lnk schema ->
+          match !lnk with
+          | Serialized _ | Small _ -> ()
+          | In_memory v -> write_child fd lnk schema v
+          | On_disk _ -> write_child fd lnk schema (fetch lnk) ) }
+  and write_child : type a. out_channel -> a link -> a schema -> a -> unit =
+   fun fd lnk schema v ->
+    schema iter v;
+    if Stdlib.Obj.(reachable_words (repr v)) > 500 then (
```

Comment:
I got the point, and I agree. Is it efficient to have a side `hashtabl` or another data structure to store child weight after serialization (or not), and to retrieve them in a way to do only the addition when the father occurs?

Reply:
The hashtbl would also enable us to de-duplicate, but I tested this a bit and it's very expensive... I was hoping that we could do something simpler:
- Now that you have fixed the map schema, the call to `schema iter v` will call `iter.yield` on every child link of `v`. This will recursively compute the size of every child link (and maybe marshal the big ones). Instead of forgetting the Small children size, we could sum them into a reference such that their parent knows their total children size.
- To speed up the parent `reachable_words` computation, we'll need to avoid it recursing over the children (since we already accumulated their total size): This can be done if we can "hide" the `Small` children, by temporarily replacing them with a `Placeholder` value. Once the parent size has been computed, we'll need to restore the `Placeholder` back to their original `Small` value.